### ISSUE TITLE
Code Python ET7 pour calculer les poids et biais 

### ISSUE THREAD

**Issue by angv5845** [angv5845/Projet_IA_calcul_embarque#10]
# ET7.py
import re
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import (Input, Conv2D, BatchNormalization,
                                     MaxPool2D, Dropout, Flatten,
                                     Dense)
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.callbacks import EarlyStopping

def parse_c_array(filename, name, shape):
    """
    Extrait le tableau `name` de dimension `shape` depuis un fichier C.
    Retourne un numpy array de forme `shape`.
    """
    text = open(filename, 'r', encoding='utf-8').read()
    # localiser la déclaration
    m = re.search(
        rf"const\s+float\s+{name}\s*\[.*?\]\s*=\s*\{{(.*?)\}};",
        text, re.S)
    if not m:
        raise ValueError(f"{name} non trouvé dans {filename}")
    body = m.group(1)
    # éliminer les commentaires et retours à la ligne
    nums = re.findall(r"[-+]?\d*\.\d+|\d+", body)
    arr = np.array(list(map(float, nums)))
    expected = np.prod(shape)
    if arr.size < expected:
        print(f"⚠️ trouvé {arr.size} valeurs pour '{name}', attendu {expected}. Padding 0.0f")
        arr = np.pad(arr, (0, expected - arr.size), constant_values=0.0)
    elif arr.size > expected:
        arr = arr[:expected]
    return arr.reshape(shape)

def build_model(input_shape=(48,13,1)):
    model = Sequential([
        Input(shape=input_shape),

        Conv2D(16, (3,3), padding='same', activation='relu'),
        BatchNormalization(),
        MaxPool2D((2,2)),

        Conv2D(32, (3,3), padding='same', activation='relu'),
        BatchNormalization(),
        MaxPool2D((2,2)),

        Dropout(0.3),

        Flatten(),
        Dense(32, activation='relu'),
        Dropout(0.3),
        Dense(1, activation='sigmoid')
    ])
    model.compile(
        optimizer=Adam(1e-3),
        loss='binary_crossentropy',
        metrics=['mse']
    )
    return model

def export_layer_weights(layer, name_w, name_b):
    """
    Affiche en C les tableaux de poids et biais d'une couche Keras.
    """
    w, b = layer.get_weights()
    flat_w = w.flatten()
    print(f"float {name_w}[] = {{")
    for v in flat_w:
        print(f"  {v:+.6f}f,")
    print("};\n")
    print(f"float {name_b}[] = {{")
    for v in b:
        print(f"  {v:+.6f}f,")
    print("};\n")

def main():
    # 1) Charger et parser les données
    X_train = parse_c_array("training.txt", "inputs", (100,48,13))
    X_test  = parse_c_array("test.txt",     "testData", (10,48,13))

    # 2) Générer y (0=Rouge pour les 50 premiers, 1=Bleu pour les 50 suivants)
    y_train = np.array([0]*50 + [1]*50)
    y_test  = np.array([0]*5  + [1]*5)

    # 3) Reshape pour Keras (ajouter le canal “1”)
    X_train = X_train[...,np.newaxis]
    X_test  = X_test[...,np.newaxis]

    # 4) Construire et entraîner
    model = build_model(input_shape=(48,13,1))
    es = EarlyStopping(monitor="val_mse", patience=20, restore_best_weights=True)
    history = model.fit(
        X_train, y_train,
        validation_data=(X_test, y_test),
        epochs=200,
        batch_size=16,
        callbacks=[es],
        verbose=2
    )

    # 5) Évaluer
    test_mse = model.evaluate(X_test, y_test, verbose=0)[1]
    print(f"\nMSE test = {test_mse:.4f}")
    if test_mse < 0.05:
        print("✅ Critère ET7 validé !")
    else:
        print("❌ MSE ≥ 0.05 — ajuster l’architecture ou hyperparamètres.")

    # 6) Exporter les poids/biais
    print("\n// === Poids et biais (C arrays) à coller dans l’Arduino IDE ===\n")
    # Conv1:
    export_layer_weights(model.layers[0], "conv1_w", "conv1_b")
    # Conv2:
    export_layer_weights(model.layers[3], "conv2_w", "conv2_b")
    # Dense1:
    export_layer_weights(model.layers[7], "dense1_w", "dense1_b")
    # Dense2:
    export_layer_weights(model.layers[9], "dense2_w", "dense2_b")

if __name__ == "__main__":
    main()

**Comment by angv5845** [comment]
import re
import numpy as np
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.callbacks import EarlyStopping

def parse_c_array(filename, name, shape):
    """Lit un fichier C contenant 'const float name[...] = { ... };'
       et renvoie un numpy array de la forme voulue."""
    text = open(filename, 'r', encoding='utf-8').read()
    # Extraire tout ce qui se trouve entre name[...] = { et le } final
    m = re.search(rf'{re.escape(name)}\s*\[.*?\]\s*=\s*\{{(.+?)\}}\s*;', text, re.S)
    if not m:
        raise ValueError(f"{name} non trouvé dans {filename}")
    # Récupérer tous les floats
    vals = re.findall(r'[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?\d+)?', m.group(1))
    arr = np.array([float(v) for v in vals], dtype=np.float32)
    expected = np.prod(shape)
    if arr.size < expected:
        print(f"⚠️  trouvé {arr.size} valeurs pour '{name}', padding 0.")
        arr = np.pad(arr, (0, expected - arr.size), 'constant')
    elif arr.size > expected:
        print(f"⚠️  trouvé {arr.size} valeurs pour '{name}', tronqué.")
        arr = arr[:expected]
    return arr.reshape(shape)

def build_model():
    inp = keras.Input(shape=(48, 13, 1))
    x = layers.Conv2D(8, 3, padding='same', activation='relu', name='conv1')(inp)
    x = layers.MaxPool2D(2, name='pool1')(x)
    x = layers.Conv2D(16, 3, padding='same', activation='relu', name='conv2')(x)
    x = layers.MaxPool2D(2, name='pool2')(x)
    x = layers.Flatten(name='flatten')(x)
    x = layers.Dense(32, activation='relu', name='dense1')(x)
    out = layers.Dense(1, activation='sigmoid', name='dense2')(x)
    model = keras.Model(inp, out, name='cnn_mlp')
    model.compile(optimizer='adam', loss='mse', metrics=['mse'])
    return model

def export_layer_weights(layer, prefix, file_handle):
    """Si la couche a des poids, les écrit sous forme C dans le fichier."""
    w = layer.get_weights()
    if not w:
        return
    kernel, bias = w
    flat_k = kernel.flatten()
    flat_b = bias.flatten()
    file_handle.write(f"// --- {layer.name} weights & bias ---\n")
    file_handle.write(f"const float {prefix}_w[] = {{\n")
    for v in flat_k:
        file_handle.write(f"  {v:+.6f}f,\n")
    file_handle.write("};\n\n")
    file_handle.write(f"const float {prefix}_b[] = {{\n")
    for v in flat_b:
        file_handle.write(f"  {v:+.6f}f,\n")
    file_handle.write("};\n\n")

def main():
    # 1) Chargement des données
    X_train = parse_c_array("training.txt", "inputs",  (100,48,13))
    y_train = np.array([0]*50 + [1]*50, dtype=np.float32)
    X_test  = parse_c_array("test.txt",     "testData",(10,48,13))
    y_test  = np.array([0]*5 + [1]*5,       dtype=np.float32)

    # Reshape pour Keras (batch, height, width, channels)
    X_train = X_train[..., np.newaxis]
    X_test  = X_test[...,  np.newaxis]

    # 2) Construction et entraînement
    model = build_model()
    es = EarlyStopping(monitor="val_mse", patience=10, restore_best_weights=True)
    model.fit(
        X_train, y_train,
        validation_data=(X_test, y_test),
        epochs=200,
        batch_size=16,
        callbacks=[es],
        verbose=2
    )

    # 3) Évaluation finale
    loss, mse = model.evaluate(X_test, y_test, verbose=0)
    print(f"\nMSE test = {mse:.6f}")
    if mse < 0.05:
        print("✅ Critère ET7 validé !")
    else:
        print("❌ MSE ≥ 0.05 — à améliorer.")

    # 4) Export des poids/biais dans ET7.txt
    with open("ET7.txt", "w", encoding="utf-8") as f:
        f.write("// Poids et biais exportés pour Arduino INO\n\n")
        for i, layer in enumerate(model.layers):
            export_layer_weights(layer, f"layer{i}", f)

if __name__ == "__main__":
    main()
